File: prolog/knapsack/knapsack.pl
:- use_module(library(clpfd)).

maximum_value(Items, Capacity, Value) :-
    \+ (member(item(Weight, _), Items), Weight > Capacity),
    length(Items, N),
    length(Selected, N),
    Selected ins 0..1,

    maplist(value, Items, Values),
    maplist(weight, Items, Weights),

    scalar_product(Weights, Selected, #=<, Capacity),
    scalar_product(Values, Selected, #=, Value),
    
    labeling([max(Value)], Selected),
    !.
    
    
weight(item(Weight, _), Weight).

value(item(_, Value), Value).

File: java/knapsack/src/main/java/Knapsack.java
import java.util.List;

class Knapsack {

    int maximumValue(int maximumWeight, List<Item> items) {
        return knapsack(items.size() - 1, maximumWeight, items);
    }

    int knapsack(int idx, int cap, List<Item> items) {
        if (idx == -1 || cap == 0) return 0;
        Item it = items.get(idx);
        if (it.weight > cap) return knapsack(idx - 1, cap, items);
        int x = knapsack(idx - 1, cap, items);
        int y = it.value + knapsack(idx - 1, cap - it.weight, items);
        return Math.max(x, y);
    }

}